src/transformation-components/GenericFold.tsx

```diff
@@ -0,0 +1,185 @@
+import React, { useCallback, ReactElement, useState } from "react";
+import { getContextAndDataSet } from "../utils/codapPhone";
+import { useInput } from "../utils/hooks";
+import { TransformationProps } from "./types";
+import { DataSet } from "../transformations/types";
+import {
+  TransformationSubmitButtons,
+  ContextSelector,
+  AttributeSelector,
+  CodapFlowTextInput,
+  ExpressionEditor,
+} from "../ui-components";
+import { applyNewDataSet, readableName, addUpdateListener } from "./util";
+import { genericFold } from "../transformations/fold";
+import TransformationSaveButton from "../ui-components/TransformationSaveButton";
+
+export interface GenericFoldSaveData {
+  inputAttributeName: string | null;
+  outputAttributeName: string;
+  expression: string;
+  base: string;
+  accumulatorName: string;
+}
+
+interface GenericFoldProps extends TransformationProps {
+  saveData?: GenericFoldSaveData;
+}
+
+export function GenericFold({
+  setErrMsg,
+  saveData,
+  errorDisplay,
+}: GenericFoldProps): ReactElement {
+  const [inputDataCtxt, inputChange] = useInput<
+    string | null,
+    HTMLSelectElement
+  >(null, () => setErrMsg(null));
+
+  const [inputAttributeName, inputAttributeNameChange] = useState<
+    string | null
+  >(saveData !== undefined ? saveData.inputAttributeName : null);
+  const [outputAttributeName, outputAttributeNameChange] = useInput<
+    string,
+    HTMLInputElement
+  >(saveData !== undefined ? saveData.outputAttributeName : "", () =>
+    setErrMsg(null)
+  );
+  const [base, baseChange] = useInput<string, HTMLInputElement>(
+    saveData !== undefined ? saveData.base : "",
+    () => setErrMsg(null)
+  );
+  const [expression, expressionChange] = useState<string>(
+    saveData !== undefined ? saveData.expression : ""
+  );
+  const [accumulatorName, accumulatorNameChange] = useInput<
+    string,
+    HTMLInputElement
+  >(saveData !== undefined ? saveData.accumulatorName : "", () =>
+    setErrMsg(null)
+  );
+
+  const transform = useCallback(async () => {
+    setErrMsg(null);
+
+    if (inputDataCtxt === null) {
+      setErrMsg("Please choose a valid dataset to transform.");
+      return;
+    }
+    if (inputAttributeName === null) {
+      setErrMsg("Please select an attribute to aggregate");
+      return;
+    }
+    if (outputAttributeName === "") {
+      setErrMsg("Please enter a name for the new attribute");
+      return;
+    }
+    if (expression === "") {
+      setErrMsg("Please enter an expression");
+      return;
+    }
+    if (base === "") {
+      setErrMsg("Please enter a base value");
+      return;
+    }
```

Comment:
Hmm, if I'm doing a fold over strings and want the base value to be the empty string, this disallows that.

This makes me wonder if what they input for the base value should actually be a full expression that we evaluate to get the base value. This seems like it might be useful and would get around the above issue. Thoughts?

Reply:
I think that's a good idea. Then empty strings will be more explicit.